src/cache.rs

```diff
@@ -0,0 +1,292 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+use std::cmp::Ord;
+use std::fmt::Debug;
+
+use rusqlite;
+
+use errors::*;
+use mentat_core::{
+    HasSchema,
+    KnownEntid,
+    NamespacedKeyword,
+    Schema,
+    TypedValue,
+};
+
+use query::{
+    q_once,
+    IntoResult,
+    QueryInputs,
+    Variable,
+};
+
+pub enum CacheAction {
+    Register,
+    Deregister,
+}
+
+pub trait ValueProvider<K, V>: Sized {
+    fn fetch_values<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema) -> Result<BTreeMap<K, V>>;
+}
+
+pub trait Cacheable {
+    type Key;
+    type Value;
+    type ValueProvider;
+
+    fn new(value_provider: Self::ValueProvider) -> Self;
+    fn cache_values<'sqlite, 'schema>(&mut self,
+                             sqlite: &'sqlite rusqlite::Connection,
+                             schema: &'schema Schema) -> Result<()>;
+    fn update_values<'sqlite, 'schema>(&mut self,
+                                       sqlite: &'sqlite rusqlite::Connection,
+                                       schema: &'schema Schema) -> Result<()>;
+    fn get(&self, key: &Self::Key) -> Option<&Self::Value>;
+}
+
+pub struct EagerCache<K, V, VP> where K: Ord, VP: ValueProvider<K, V> {
+    cache: BTreeMap<K, V>,
+    value_provider: VP,
+}
+
+impl<K, V, VP> Cacheable for EagerCache<K, V, VP> where K: Ord + Clone + Debug, V: Clone, VP: ValueProvider<K, V> {
+    type Key = K;
+    type Value = V;
+    type ValueProvider = VP;
+
+    fn new(value_provider: Self::ValueProvider) -> Self {
+        EagerCache {
+            cache: BTreeMap::new(),
+            value_provider,
+        }
+    }
+
+    fn cache_values<'sqlite, 'schema>(&mut self,
+                                            sqlite: &'sqlite rusqlite::Connection,
+                                            schema: &'schema Schema) -> Result<()> {
+        // fetch results and add to cache
+        self.cache = self.value_provider.fetch_values(sqlite, schema)?;
+        Ok(())
+    }
+
+    fn update_values<'sqlite, 'schema>(&mut self,
+                                      sqlite: &'sqlite rusqlite::Connection,
+                                      schema: &'schema Schema) -> Result<()> {
+        self.cache_values(sqlite, schema)
+    }
+
+    fn get(&self, key: &Self::Key) -> Option<&Self::Value> {
+        self.cache.get(&key)
+    }
+}
+
+pub struct AttributeCacher {
+    cache: BTreeMap<NamespacedKeyword, EagerCache<TypedValue, Vec<TypedValue>, AttributeValueProvider>>,   // values keyed by attribute
```

Comment:
Cache based on `Entid`, not `NamespacedKeyword`, and on `Entid`, not `TypedValue`, for the left side of the cache.

Consider that there are two kinds of attributes — ones with zero or one value (cardinality one), and one with zero or more values (cardinality many). Skipping the `Vec` in the former case is convenient…